segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorageTest.java

```diff
@@ -0,0 +1,282 @@
+                    ex -> ex instanceof IllegalArgumentException);
+
+            assertThrows(
+                    "write() did not throw for handle pointing to inexistent segment.",
+                    () -> s.write(createHandle(segmentName + "_1", false, DEFAULT_EPOCH), 0, new ByteArrayInputStream("h".getBytes()), 1, TIMEOUT),
+                    ex -> ex instanceof StreamSegmentNotExistsException);
+
+            val writeHandle = s.openWrite(segmentName).join();
+            long offset = 0;
+            for (int j = 0; j < appendCount; j++) {
+                byte[] writeData = String.format("Segment_%s_Append_%d", segmentName, j).getBytes();
+                ByteArrayInputStream dataStream = new ByteArrayInputStream(writeData);
+                s.write(writeHandle, offset, dataStream, writeData.length, TIMEOUT).join();
+                offset += writeData.length;
+            }
+
+            // Check bad offset.
+            final long finalOffset = offset;
+            assertThrows("write() did not throw bad offset write (larger).",
+                    () -> s.write(writeHandle, finalOffset + 1, new ByteArrayInputStream("h".getBytes()), 1, TIMEOUT),
+                    ex -> ex instanceof BadOffsetException);
+
+            // Check post-delete write.
+            s.delete(writeHandle, TIMEOUT).join();
+            assertThrows("write() did not throw for a deleted StreamSegment.",
+                    () -> s.write(writeHandle, 0, new ByteArrayInputStream(new byte[1]), 1, TIMEOUT),
+                    ex -> ex instanceof StreamSegmentNotExistsException);
+        }
+    }
+
+    //endregion
+
+    //region synchronization unit tests
+
+    /**
+     * This test case simulates two hosts writing at the same offset at the same time.
+     */
+    @Test
+    public void testParallelWriteTwoHosts() {
+        String segmentName = "foo_write";
+        int appendCount = 5;
+
+        try (Storage s1 = createStorage();
+             Storage s2 = createStorage()) {
+            s1.initialize(DEFAULT_EPOCH);
+            s1.create(segmentName, TIMEOUT).join();
+            SegmentHandle writeHandle1 = s1.openWrite(segmentName).join();
+            SegmentHandle writeHandle2 = s2.openWrite(segmentName).join();
+            long offset = 0;
+            byte[] writeData = String.format("Segment_%s_Append", segmentName).getBytes();
+            for (int j = 0; j < appendCount; j++) {
+                ByteArrayInputStream dataStream1 = new ByteArrayInputStream(writeData);
+                ByteArrayInputStream dataStream2 = new ByteArrayInputStream(writeData);
+                CompletableFuture f1 = s1.write(writeHandle1, offset, dataStream1, writeData.length, TIMEOUT);
+                CompletableFuture f2 = s2.write(writeHandle2, offset, dataStream2, writeData.length, TIMEOUT);
+                assertMayThrow("Write expected to complete OR throw BadOffsetException." +
```

Comment:
I'm a bit worried about this use of `assertMayThrow` because the behavior isn't well defined. Is it fine if it never throws? Is it fine if it always throws?